Plugins/Public/mobiledocking_plugin/Main.cpp

```diff
@@ -201,252 +82,618 @@ void LoadSettings()
-		PrintUserCmdText(client, L"Navmap updated successfully.");
+			pub::SpaceObj::GetLocation(carrierShipID, pos, ori);
+			pub::SpaceObj::GetSystem(carrierShipID, iSystemID);
 
-		// Update current system stat in player list to be displayed relevantly.
-		Server.BaseEnter(mobiledockClients[client].proxyBaseID, client);
-		Server.BaseExit(mobiledockClients[client].proxyBaseID, client);
-		wstring wscCharFileName;
-		HkGetCharFileName((const wchar_t*)Players.GetActiveCharacterName(client), wscCharFileName);
-		wscCharFileName += L".fl";
-		CHARACTER_ID cID;
-		strcpy(cID.szCharFilename, wstos(wscCharFileName.substr(0, 14)).c_str());
-		Server.CharacterSelect(cID, client);
+			pos.x += ori.data[0][1] * set_iMobileDockOffset;
+			pos.y += ori.data[1][1] * set_iMobileDockOffset;
+			pos.z += ori.data[2][1] * set_iMobileDockOffset;
+
+			jumpData.iClientID = client;
+			jumpData.iSystemID = iSystemID;
+			jumpData.pos = pos;
+			jumpData.ori = ori;
+
+			Plugin_Communication(PLUGIN_MESSAGE::CUSTOM_JUMP_CALLOUT, &jumpData);
+			Players[client].iLastBaseID = idToDockedInfoMap[client]->lastDockedSolar;
+		}
 
-		// Update current system stat in plugin data.
-		Universe::IBase* base = Universe::get_base(mobiledockClients[client].proxyBaseID);
-		mobiledockClients[client].carrierSystem = base->iSystemID;
+		mobiledockClients[carrierClientID].iDockingModulesAvailable++;
+	}
+	else
+	{
+		CUSTOM_BASE_LAST_DOCKED_STRUCT lastDockedCheck;
+		lastDockedCheck.iClientID = client;
+		lastDockedCheck.iLastDockedBaseID = 0;
+		Plugin_Communication(PLUGIN_MESSAGE::CUSTOM_BASE_LAST_DOCKED, &lastDockedCheck);
+		if (lastDockedCheck.iLastDockedBaseID)
+		{
+			CUSTOM_BASE_BEAM_STRUCT POBbeamStruct;
+			POBbeamStruct.iClientID = client;
+			POBbeamStruct.iTargetBaseID = lastDockedCheck.iLastDockedBaseID;
+			Plugin_Communication(PLUGIN_MESSAGE::CUSTOM_BASE_BEAM, &POBbeamStruct);
+		}
+		else
+		{
+			pub::Player::ForceLand(client, idToDockedInfoMap[client]->lastDockedSolar);
+		}
 	}
+
+	wstring dockedCharname = (const wchar_t*)Players.GetActiveCharacterName(client);
+	RemoveShipFromLists(dockedCharname, false);
 }
 
 // If this is a docking request at a player ship then process it.
 int __cdecl Dock_Call(unsigned int const &iShip, unsigned int const &iBaseID, int iCancel, enum DOCK_HOST_RESPONSE response)
 {
 	returncode = DEFAULT_RETURNCODE;
 
-	UINT client = HkGetClientIDByShip(iShip);
+	//if not a player dock, skip
+	uint client = HkGetClientIDByShip(iShip);
 	if (client)
 	{
-		// If no target then ignore the request.
-		uint iTargetShip;
-		pub::SpaceObj::GetTarget(iShip, iTargetShip);
-		if (!iTargetShip)
-			return 0;
-
+		// If target not a player in FREIGHTER class ship, ignore request
 		uint iType;
-		pub::SpaceObj::GetType(iTargetShip, iType);
+		pub::SpaceObj::GetType(iBaseID, iType);
 		if (iType != OBJ_FREIGHTER)

```

Comment:
With the new patch you should check CAPITAL, CRUISER, etc

Reply:
Adjusted in a separate PR that globally replaced FREIGHTER with CRUISER | CAPITAL